src/slice-array.c

```diff
@@ -143,66 +57,56 @@ struct vec_slice_shaped_info {
-      *out_data = x_data[loc];                                                        \
-    }                                                                                 \
-                                                                                      \
-    /* Update shape_index */                                                          \
-    for (int j = 0; j < info.shape_n; ++j) {                                          \
-      info.p_shape_index[j]++;                                                        \
-      if (info.p_shape_index[j] < info.p_dim[j + 1]) {                                \
-        break;                                                                        \
-      }                                                                               \
-      info.p_shape_index[j] = 0;                                                      \
-    }                                                                                 \
-  }                                                                                   \
-                                                                                      \
-  UNPROTECT(1);                                                                       \
+#define SLICE_SHAPED_COMPACT_SEQ(RTYPE, CTYPE, DEREF, CONST_DEREF) \
+  SEXP out_dim = PROTECT(Rf_shallow_duplicate(p_info->dim));       \
+  INTEGER(out_dim)[0] = p_info->index_n;                           \
+                                                                   \
+  SEXP out = PROTECT(Rf_allocArray(RTYPE, out_dim));               \
+  CTYPE* out_data = DEREF(out);                                    \
+                                                                   \
+  R_len_t start = p_info->p_index[0];                              \
+  R_len_t n = p_info->p_index[1];                                  \
+  R_len_t step = p_info->p_index[2];                               \
+                                                                   \
+  const CTYPE* x_data = CONST_DEREF(x);                            \
+                                                                   \
+  for (int i = 0; i < p_info->shape_elem_n; ++i) {                 \
+    int loc = vec_strided_loc(                                     \
+      p_info->p_shape_index,                                       \
+      p_info->p_strides,                                           \
+      p_info->shape_n                                              \
+    );                                                             \
+                                                                   \
+    loc += start;                                                  \
+                                                                   \
+    for (int j = 0; j < n; ++j, ++out_data, loc += step) {         \
+      *out_data = x_data[loc];                                     \
```

Comment:
Now I'm wondering if `loc` should be a `R_len_t`, since we use it to index the data?